### ISSUE TITLE
LoRA training runs the whole transformer in float32, whatever precision the model was loaded in

### ISSUE THREAD

**Issue by azrahello** [mflux-community/mflux#740]
`LoRALinear.__call__` multiplies the float32 adapter factors straight against the activation:

  ```python
  # src/mflux/models/common/lora/layer/linear_lora_layer.py
  lora_out = mx.matmul(mx.matmul(x, self.lora_A), self.lora_B)
  return base_out + self.scale * lora_out

  ModelConfig.precision is bfloat16, so MLX promotes bf16 @ fp32 to fp32. base_out + scale * lora_out is then
  fp32 and feeds the next block, which promotes again. With a few
  hundred patched layers the whole transformer ends up running in float32 — double the
  activation memory and bandwidth for a model that was loaded in half precision.

  The same applies to FusedLoRALinear (stacked adapters) and LoKrLinear.

  1. Train a LoRA on any model, dense, with enough targets to cover the transformer.
  2. The activations are float32 throughout, not ModelConfig.precision.

  Measured on an M2 Ultra 64 GB, Krea 2 training, 264 targets at rank 16, max_resolution 512:

  ┌────────────────────────────────────┬────────────┬───────────────────────────────────────────────────────┐
  │                                    │ step time  │                        memory                         │
  ├────────────────────────────────────┼────────────┼───────────────────────────────────────────────────────┤
  │ dense, un-cast (float32            │ 74 s/step  │ swap saturated, kern.memorystatus_vm_pressure_level = │
  │ activations)                       │            │  2                                                    │
  ├────────────────────────────────────┼────────────┼───────────────────────────────────────────────────────┤
  │ dense, cast to activation dtype    │ 9.5 s/step │ pressure level 1                                      │
  ├────────────────────────────────────┼────────────┼───────────────────────────────────────────────────────┤
  │ q8 (the workaround), cast          │ 13.0       │ —                                                     │
  │                                    │ s/step     │                                                       │
  └────────────────────────────────────┴────────────┴───────────────────────────────────────────────────────┘

  The cost is not only speed. On a machine where the dense model would otherwise fit, it forces
  quantization — and quantizing then costs a further 27% per step for nothing: validation loss
  over 200 steps is the same either way (dense 0.1221 → 0.0961, q8 0.1225 → 0.0970).

  The factors themselves should stay float32 as master weights (fp16 factors underflow the
  gradient at typical learning rates), so the cast belongs inside the forward, not at init.
  Casting an (in_dims, r) matrix is free next to the base GEMM.

  This does not lower the precision the optimizer sees. Measured, with and without the cast:

  ┌──────────────────────────────────┬───────────────────────────────────────┐
  │                                  │                                       │
  ├──────────────────────────────────┼───────────────────────────────────────┤
  │ gradient dtype                   │ float32 in both cases                 │
  ├──────────────────────────────────┼───────────────────────────────────────┤
  │ relative difference in gradients │ 0.26% (ordinary half-precision noise) │
  ├──────────────────────────────────┼───────────────────────────────────────┤
  │ float32 activations              │ cast is a no-op                       │
  └──────────────────────────────────┴───────────────────────────────────────┘

  To check without a training run:

  import mlx.core as mx
  from mlx import nn
  from mflux.models.common.lora.layer.linear_lora_layer import LoRALinear

  linear = nn.Linear(64, 64, bias=False)
  linear.weight = linear.weight.astype(mx.bfloat16)
  layer = LoRALinear.from_linear(linear, r=8)
  print(layer(mx.zeros((1, 4, 64), mx.bfloat16)).dtype)   # float32; should be bfloat16

  Fix: cast the factors to x.dtype inside the forward in linear_lora_layer.py,
  fused_linear_lora_layer.py and linear_lokr_layer.py. The master weights are untouched.
